fastavro/_writer.pyx

```diff
@@ -449,9 +451,9 @@ cpdef write_union(bytearray fo, datum, schema):
             raise ValueError(msg)
     else:
         pytype = type(datum)
-        for index, candidate in enumerate(schema):
-            if validate(datum, candidate):
-                break
+        validation_scores = [validate(datum, candidate) for candidate in schema]
```

Comment:
Do you think we could choose the highest-scoring option without building and scanning a temporary list? It could be significantly faster that way. Maybe something like this:
```
cdef int validation_score
cdef int best_validation_score = -1
cdef int i, index, best_index, validation_score
for i, candidate in enumerate(schema):
    validation_score = validate(datum, candidate)
    if validation_score > best_validation_score:
        best_validation_score = validation_score
        best_index = i
if best_validation_score > 0:
    index = best_index
else:
    msg = '%r (type %s) do not match %s' % (datum, pytype, schema)
    raise ValueError(msg)
```

Reply:
Thanks for the feedback. I changed it in a way https://github.com/tebeka/fastavro/pull/129#issuecomment-350354076 was suggesting and it includes your remark too.